secure/storage/src/storage.rs

```diff
@@ -47,33 +50,137 @@ pub trait Storage: Send + Sync {
-    /// given 'key_pair_name'. If the key pair doesn't exist, or the caller doesn't have the
+    /// Returns the public key for a given Ed25519 key pair, as identified by the 'key_pair_name'.
+    /// If the key pair doesn't exist, or the caller doesn't have the
     /// appropriate permissions to retrieve the public key, this call will fail with an error.
-    fn get_public_key_for(&mut self, key_pair_name: &str) -> Result<Ed25519PublicKey, Error> {
+    fn get_public_key_for_name(&self, key_pair_name: &str) -> Result<Ed25519PublicKey, Error> {
+        self.get_private_key_for_name(key_pair_name)
+            .map(|e| e.public_key())
+    }
+
+    /// Returns the private key for a given Ed25519 key pair, as identified by the 'key_pair_name'.
+    /// If the key pair doesn't exist, or the caller doesn't have the appropriate permissions to
+    /// retrieve the private key, this call will fail with an error.
+    fn get_private_key_for_name(&self, key_pair_name: &str) -> Result<Ed25519PrivateKey, Error> {
+        match self.get(key_pair_name)? {
+            Value::Ed25519PrivateKey(private_key) => Ok(private_key),
+            _ => Err(Error::UnexpectedValueType),
+        }
+    }
+
+    /// Returns the private key for a given Ed25519 key pair version, as identified by the
+    /// 'key_pair_name' and 'key_pair_version'. If the key pair at the specified version doesn't
+    /// exist, or the caller doesn't have the appropriate permissions to retrieve the private key,
+    /// this call will fail with an error.
+    fn get_private_key_for_name_and_version(
+        &self,
+        key_pair_name: &str,
+        key_pair_version: Ed25519PublicKey,
+    ) -> Result<Ed25519PrivateKey, Error> {
+        let current_private_key = self.get_private_key_for_name(key_pair_name)?;
+        if current_private_key.public_key().eq(&key_pair_version) {
+            return Ok(current_private_key)
+        }
+
+        let previous_private_key =
+            self.get_private_key_for_name(&get_previous_version_name(key_pair_name))?;
+        if previous_private_key.public_key().eq(&key_pair_version) {
+            return Ok(previous_private_key)
+        }
+
+        Err(Error::KeyVersionNotFound(key_pair_version.to_string()))
+    }
+
+    /// Rotates an Ed25519 key pair by generating a new Ed25519 key pair, and updating the
+    /// 'key_pair_name' to reference the freshly generated key. The previous key pair is retained
+    /// in storage if needed. If multiple key rotations occur over the lifetime of a key pair, only
+    /// two versions of the key pair are maintained (i.e., the current and previous one).
+    /// If the key pair doesn't exist, or the caller doesn't have the appropriate permissions to
+    /// retrieve the public key, this call will fail with an error. Otherwise, the new public
+    /// key for the rotated key pair is returned.
+    fn rotate_key_pair(&mut self, key_pair_name: &str) -> Result<Ed25519PublicKey, Error> {
```

Comment:
how do you clean up older vault key versions? even though we limit the number of keys in the system, what are we doing to limit the versions of those keys?

Reply:
Thanks David!
1. In regards to your other comment (the one I resolved -- sorry :P), the purpose of this PR is to find a simple way of supporting key rotation and versioning that is compatible with all backends. While Transit is the main goal for the Vault backend, trying to leverage it in this PR will blow the PR up significantly and make the changes difficult to reason about. Instead, this PR provides an MVP that is compatible with the vault K/V store. Moreover, there's nothing here that blocks me from using Transit in a future PR (which is the plan -- see the issue description). 
2. In regards to cleaning up vault key versions, this is indeed an issue (hence why I've chosen the approach here as a first step that works with vault). However, from what I understand, when we come to do the Transit specific implementation for Vault, "Trim Key" operation should support this  functionality (https://www.vaultproject.io/api/secret/transit/index.html#trim-key)?

Reply:
I don't disagree but I think exposing the "general" versioning scheme in the secure storage trait makes it look like it is the expected and only implementation and makes it harder to move away from if we want leverage native versioning. Because you'll be first having to refactor your interface.

I can push on this topic later, but in general I prefer a model where we minimize introductions into that layer unless we are confident it generalizes across all platforms or it can be hidden by monomorphic behaviors

@bors-libra delegate+